### PR TITLE
[SPARK-9350][SQL] Introduce an InternalRow generic getter that requires a DataType

### PR DESCRIPTION
**Pull request by rxin**
Currently UnsafeRow cannot support a generic getter. However, if the data type is known, we can support a generic getter.

### CODE REVIEW

**Comment by yhuai** [comment]
Aggregate function part looks good.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala [hunk 1]
```diff
@@ -32,32 +32,36 @@ abstract class InternalRow extends Serializable {
 
   def get(ordinal: Int): Any
 
-  def getAs[T](ordinal: Int): T = get(ordinal).asInstanceOf[T]
+  def genericGet(ordinal: Int): Any = get(ordinal, null)
+
+  def get(ordinal: Int, dataType: DataType): Any = get(ordinal)
+
+  def getAs[T](ordinal: Int, dataType: DataType): T = get(ordinal, dataType).asInstanceOf[T]
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala]: where do we use the `dateType`?

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala]: It's not implemented yet, but in UnsafeRow.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala [hunk 2]
```diff
@@ -32,32 +32,36 @@ abstract class InternalRow extends Serializable {
 
   def get(ordinal: Int): Any
 
-  def getAs[T](ordinal: Int): T = get(ordinal).asInstanceOf[T]
+  def genericGet(ordinal: Int): Any = get(ordinal, null)
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala]: This is created as a special marker. For places that we currently use get without a data type that will be gone but I don't have time to delete yet (or the ones that don't matter), I will replace them with this genericGet.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SpecificMutableRow.scala [hunk 3]
```diff
@@ -221,6 +221,10 @@ final class SpecificMutableRow(val values: Array[MutableValue]) extends MutableR
 
   override def get(i: Int): Any = values(i).boxed
 
+  override def getStruct(ordinal: Int, numFields: Int): InternalRow = {
+    values(ordinal).boxed.asInstanceOf[InternalRow]
+  }
+
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SpecificMutableRow.scala]: why override `getStruct` here? It seems the same with default implementation of `getStruct`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/rows.scala [hunk 4]
```diff
@@ -101,6 +101,10 @@ class GenericInternalRow(protected[sql] val values: Array[Any]) extends Internal
 
   override def get(i: Int): Any = values(i)
 
+  override def getStruct(ordinal: Int, numFields: Int): InternalRow = {
+    values(ordinal).asInstanceOf[InternalRow]
+  }
+
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/rows.scala]: why override?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/rows.scala [hunk 5]
```diff
@@ -128,6 +132,10 @@ class GenericMutableRow(val values: Array[Any]) extends MutableRow {
...
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/rows.scala]: why override?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala (hunk 2 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala]: do you mean we will remove `get(ordinal: Int)` in the future?

**Comment by cloud-fan** [comment]
LGTM

**Comment by rxin** [comment]
Thanks - I'm going to merge this and submit a followup pr with more stuff.

**Pull request closed without merging**